fish/fish.6
.\" This file is free software, distributed under the BSD license.
.Dd May 31, 1993
.Dt FISH 6
.Os
.Sh NAME
.Nm fish
.Nd play
.Dq Go Fish
.Sh DESCRIPTION
.Nm
is the game
.Dq Go Fish ,
a traditional children's card game.
.Pp
The computer deals the player and itself seven cards, and places
the rest of the deck face-down (figuratively).
The object of the game is to collect
.Dq books ,
or all of the members of a single rank.
For example, collecting four 2's would give the player a
.Dq book of 2's .
.Pp
The computer makes a random decision as to who gets to start the game,
and then the computer and player take turns asking each other for cards
of a specified rank. If the asked player has any cards of the requested
rank, they give them up to the asking player. A player must have at least
one of the cards of the rank they request in his hand. When a player asks
for a rank of which the other player has no cards, the asker is told to
.Dq Go Fish! .
Then, the asker draws a card from the non-dealt cards. If he draws the
card he asked for, he continue his turn, asking for more ranks from the
other player. Otherwise, the other player gets a turn.
.Pp
When a player completes a book, either by getting cards from the other
player or drawing from the deck, they set those cards aside and the rank
is no longer in play.
.Pp
The game ends when either player no longer has any cards in their hand.
The player with the most books wins.
.Pp
.Nm
provides instructions as to what input it accepts.


doc/NEWS
NEWS for bsd-games and bsd-games-non-free
=========================================

This file contains a summary of changes in each version of bsd-games
and bsd-games-non-free after 1.3.  For fuller details, see ChangeLog
and ChangeLog.0.

As well as reading this summary, it is a good idea to reread the
INSTALL file whenever installing a new version, to check for any
changes in installation instructions.  Packagers should also reread
the PACKAGING file each version (or do a diff between the old and new
versions of the package and read that).

Version 2.17 (2005-02-18)
============

* Updated from NetBSD CVS of 2005-02-18.

* Some bugs fixed (in particular, should now build with current GCC 4
CVS), some backlogged.

Version 2.16 (2004-02-13)
============

* Updated from NetBSD CVS of 2004-02-13.  Includes a new game, dab
(Dots and Boxes), written in C++.

* Supplementary documentation for trek and rogue now installed
automatically (in a configurable directory).

* Some changes for portability to Cygwin, thanks to David A Bagley
<bagleyd@tux.org>; more portability issues remain.

* Various improvements to morse from OpenBSD.

* Some bugs fixed, some backlogged.

Version 2.15 (2003-12-18)
============

* Updated from NetBSD CVS of 2003-12-17.  This includes:

  + Advertising clause removed from licence of almost all
  UCB-copyright code.

  + Hunt licence clarified.

* Various bugs fixed, though some bug reports are still backlogged.

* Glibc 2.3 or later now required, at least for hunt.

* Directory for fortune files individually configurable, and default
has changed to match that used in NetBSD.

Version 2.14 (2003-04-01)
============

* Hack is now BSD-licensed
<http://www.cwi.nl/~aeb/games/hack/hack.html>, so moves from
bsd-games-non-free to bsd-games.

* Updated from NetBSD CVS of 2003-03-29.  This includes:

  + Security fix for rogue (in bsd-games-non-free); see NetBSD
  Security Advisory 2002-021.

  + Optional support for large numbers in factor using libcrypto from
  OpenSSL.  This can handle larger numbers than GNU factor (which is
  limited to 64-bit numbers) but not that much larger efficiently, as
  the algorithm used (Pollard rho) is O(n^{1/4}) in the number
  factored.

* Some bugs fixed, though a backlog of bug reports remains.

Version 2.13 (2001-08-30)
============

* Warning now present in README about the inclusion of cryptographic
software (caesar and rot13).

* Various minor improvements and bugfixes, some from NetBSD, some from
OpenBSD, some from other contributors.  Current with NetBSD CVS of
2001-08-29.

* Some fixes to avoid some warnings when compiling with current glibc
and GCC 3.

* May support GNU Hurd (untested).

Version 2.12 (2000-09-25)
============

* Improvements and bugfixes to battlestar, mainly from OpenBSD.
Though there are still plenty of known bugs, it should be much harder
to crash.

* Should compile with glibc 2.2 (tested with 2.1.94).

* Other minor changes and bugfixes; current with NetBSD CVS of
2000-09-25.

Version 2.11 (2000-04-19)
============

* bsd-games-non-free now includes hack 1.0.3 by Andries Brouwer; this
may be of interest to nethack players wishing to play an ancestral
game.

* Various minor changes and bugfixes; current with NetBSD CVS of
2000-04-19.

Version 2.10 (2000-02-12)
============

* Avoid world-readable characs file in phantasia.  Note the new
configuration question which will need an appropriate answer,
especially if packaging for a distribution.

* Various minor bugfixes.

* Essentially all relevant changes, except to hunt (which will
eventually be replaced by the OpenBSD version), now merged back into
NetBSD.

* Some minor changes from NetBSD CVS of 2000-02-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.9 (1999-11-12)
===========

* Fixes various minor bugs reported recently.

* Some minor changes from NetBSD CVS of 1999-11-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.8 (1999-09-26)
===========

* Robots scorefile format has changed - remove or rename your old
score file when upgrading to this version.

* Various minor changes and bug-fixes (some from OpenBSD).

* Snake substantially cleaned up, and now uses the native curses
interface (rather than termcap emulation in ncurses); backgammon and
tetris still to be converted.

* New "game": wtf.

* Updated from NetBSD CVS of 1999-09-25; many more changes merged back
into NetBSD.

* All constant data files are now architecture independent (installed
under /usr/share by default).

* bsd-games is no longer tested by me with libc older than glibc
2.1.2; fixes for compilation with older libc will still be accepted.

* Details of the copyrights and licence terms applicable to these
games, as best as I can determine them, now appear in COPYING.

* bsd-games no longer comes with its own word list for boggle and
hangman; you need to provide one at compile time (for boggle) and at
run time (for hangman).  The copyright status of the one formerly
included was not entirely clear.  The word list used by hangman can
also now be specified at run time with the new "-d" option.

* bsd-games now has a testsuite (for programs that can be tested
non-interactively only).

Version 2.7 (1999-04-04)
===========

* Bug fixes.

* Updated from NetBSD-current of 1999-03-26; more changes merged back
into NetBSD.

* Other minor changes.

Version 2.6 (1998-12-20)
===========

* Buffer overrun fixes in sail (not thought exploitable).

* Fixes possible score file corruption if fds 0, 1, 2 aren't open at
game startup.

* Updated from NetBSD-current of 1998-12-07 (morse now has a decoding
option).

* Now builds cleanly with -Wwrite-strings.

* Other minor changes.

* Some users of this package may be interested in my "nppt" package,
which provides an alternative implementation of ppt(6) with additional
features such as Encapsulated PostScript output.  (This package

fish/fish.c
// This file is free software, distributed under the BSD license.

#include "../config.h"
#include <sys/wait.h>

//----------------------------------------------------------------------

enum {
    RANKS	= 13,
    SUITS	= 4,
    HANDSIZE	= 7,
    NCARDS	= RANKS * SUITS
};
enum EPlayer {
    USER,
    COMPUTER,
    NPLAYERS
};
static enum EPlayer OTHER (enum EPlayer p) { return 1-p; }

enum {
    PANEL_LINES = 4,
    PANEL_COLS = 80
};
enum {
    color_None,
    color_Panel,
    color_CardBlack,
    color_CardRed,
};

//----------------------------------------------------------------------

static const char c_CardNames [RANKS][3] =
    { "A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K" };
static const char c_PlayerName [NPLAYERS][4] = {"You", "I"};

static uint8_t	_hand [NPLAYERS][RANKS] = {{0},{0}};
static uint8_t	_deck [NCARDS] = {0};
static uint8_t	_decksz = NCARDS;
static bool	_asked [NPLAYERS][RANKS] = {{false},{false}};

static WINDOW*	_wmsg = NULL;
static WINDOW*	_wpanel = NULL;

//----------------------------------------------------------------------

static void shuffle_deck_and_deal (void);
static void initialize_windows (void);
static unsigned parse_card_name (char k);
static unsigned makemove (enum EPlayer player);
static unsigned usermove (void);
static unsigned compmove (void);
static unsigned drawcard (enum EPlayer player);
static bool gofish (enum EPlayer player, unsigned askedfor);
static void goodmove (enum EPlayer player, unsigned m);
static void chkwinner (enum EPlayer player);
static void printplayer (enum EPlayer player);
static unsigned countbooks (enum EPlayer player);
static void instructions (void);

//----------------------------------------------------------------------

int main (void)
{
    initialize_curses();
    initialize_windows();
    shuffle_deck_and_deal();
    instructions();

    enum EPlayer player = nrand(NPLAYERS);
    printplayer (player);
    waddstr (_wmsg, "get to start.\n");
    for (;;) {
	unsigned m = makemove (player);
	if (_hand[OTHER(player)][m])
	    goodmove (player, m);
	else if (!gofish (player, m))
	    player = OTHER(player);
    }
    return EXIT_SUCCESS;
}

static void initialize_windows (void)
{
    if (_wmsg)
	delwin (_wmsg);
    else {
	static const struct color_pair c_Pairs[] = {
	    { COLOR_WHITE,	COLOR_BLUE	},	// color_Panel
	    { COLOR_BLACK,	COLOR_WHITE	},	// color_CardBlack
	    { COLOR_RED,	COLOR_WHITE	}	// color_CardRed
	};
	init_pairs (ArrayBlock (c_Pairs));
    }
    _wmsg = newwin (LINES-PANEL_LINES, PANEL_COLS, 0, 0);
    scrollok (_wmsg, true);
    idlok (_wmsg, true);
    keypad (_wmsg, true);
    wmove (_wmsg, getmaxy(_wmsg)-1, 0);
    if (_wpanel)
	delwin (_wpanel);
    _wpanel = newwin (PANEL_LINES, PANEL_COLS, LINES-PANEL_LINES, 0);
    wbkgdset (_wpanel, COLOR_PAIR(color_Panel));
}

static void shuffle_deck_and_deal (void)
{
    // Shuffle deck
    for (unsigned i = 0; i < NCARDS; ++i)
	_deck[i] = i % RANKS;
    random_shuffle_u8 (_deck, NCARDS);
    // Deal HANDSIZE cards to both players
    for (unsigned i = 0; i < HANDSIZE; ++i) {
	++_hand[USER][_deck[--_decksz]];
	++_hand[COMPUTER][_deck[--_decksz]];
    }
}

static unsigned parse_card_name (char k)
{
    if (k >= 'a' && k <= 'z')
	k -= 'a'-'A';
    for (unsigned i = 0; i < ArraySize(c_CardNames); ++i)
	if (k == c_CardNames[i][0])
	    return i;
    return ArraySize(c_CardNames);
}

static unsigned makemove (enum EPlayer player)
{
    if (player == USER)
	return usermove();
    else {
	unsigned m = compmove();
	_asked[COMPUTER][m] = true;
	wattr_on (_wmsg, A_BOLD, NULL);
	wprintw (_wmsg, "I ask you for %s. ", c_CardNames[m]);
	wattr_off (_wmsg, A_BOLD, NULL);
	return m;
    }
}

static void draw_panel (void)
{
    werase (_wpanel);

    // Draw player cards and books
    mvwaddstr (_wpanel, 0, 0, "Cards: ");
    unsigned cx = strlen("Cards: "), bx = cx, nBooks = 0;
    for (unsigned i = 0; i < RANKS; ++i) {
	if (!_hand[USER][i])
	    continue;
	if (_hand[USER][i] < SUITS) {
	    for (unsigned y = 0; y < _hand[USER][i]; ++y) {
		wcolor_set (_wpanel, color_CardBlack+(y%2), NULL);
		mvwaddstr (_wpanel, y, cx, c_CardNames[i]);
	    }
	    cx += strlen(c_CardNames[i])+1;
	} else {
	    ++nBooks;
	    wcolor_set (_wpanel, color_CardBlack, NULL);
	    mvwaddstr (_wpanel, 3, bx, c_CardNames[i]);
	    bx += strlen(c_CardNames[i])+1;
	}
    }
    wattr_set (_wpanel, A_NORMAL, color_None, NULL);
    if (nBooks)
	mvwaddstr (_wpanel, 3, 0, "Books: ");

    // Draw computer cards and books
    mvwprintw (_wpanel, 0, getmaxx(_wpanel)-18, "%2u cards in deck", _decksz);
    unsigned compCards = 0, compBooks = 0;
    for (unsigned i = 0; i < RANKS; ++i) {
	if (!_hand[COMPUTER][i])		continue;
	else if (_hand[COMPUTER][i] < SUITS)	++compCards;
	else					++compBooks;
    }
    mvwprintw (_wpanel, 1, getmaxx(_wpanel)-16, "I have %u cards", compCards);
    if (compBooks)
	mvwprintw (_wpanel, 2, getmaxx(_wpanel)-16, "I have %u books", compBooks);

    wrefresh (_wpanel);
}

static unsigned usermove (void)
{
    for (;;) {
	draw_panel();
	int k = wgetch (_wmsg);
	unsigned n = parse_card_name (k);

	if (k == 'x' || k == KEY_F(10))
	    exit (EXIT_SUCCESS);
	else if (k == '?' || k == KEY_F(1))
	    instructions();
	else if (k == KEY_RESIZE)
	    initialize_windows();
	else if (n >= ArraySize(c_CardNames))
	    waddstr (_wmsg, "I don't understand!\n");
	else if (_hand[USER][n] < 1)
	    wprintw (_wmsg, "You don't have any %s's!\n", c_CardNames[n]);
	else if (_hand[USER][n] >= SUITS)
	    wprintw (_wmsg, "You have already completed a book of %s's.\n", c_CardNames[n]);
	else {
	    wprintw (_wmsg, "\nYou ask me for %s. ", c_CardNames[n]);
	    return n;
	}
    }
}

static unsigned compmove (void)
{
    for (unsigned i = 0; i < RANKS; ++i) {
	if (_asked[USER][i] && _hand[COMPUTER][i] > 0 && _hand[COMPUTER][i] < SUITS) {
	    _asked[USER][i] = false;
	    return i;
	}
    }
    if (!nrand(3)) {
	unsigned i, max;
	for (i = 0;; ++i) {
	    if (_hand[COMPUTER][i] && _hand[COMPUTER][i] != SUITS) {
		max = i;
		break;
	    }
	}
	while (++i < RANKS)
	    if (_hand[COMPUTER][i] != SUITS && _hand[COMPUTER][i] > _hand[COMPUTER][max])
		max = i;
	return max;
    }
    for (;;) {
	for (unsigned i = 0; i < RANKS; ++i)
	    if (_hand[COMPUTER][i] && _hand[COMPUTER][i] != SUITS && !_asked[COMPUTER][i])
		return i;
	for (unsigned i = 0; i < RANKS; ++i)
	    _asked[COMPUTER][i] = false;
    }
}

static unsigned drawcard (enum EPlayer player)
{
    unsigned card = _deck[--_decksz];
    ++_hand[player][card];
    if (player == USER || _hand[player][card] == SUITS) {
	printplayer(player);
	wprintw (_wmsg, "drew %s", c_CardNames[card]);
	if (_hand[player][card] == SUITS) {
	    wprintw (_wmsg, " and made a book of %s's.\n", c_CardNames[card]);
	    chkwinner (player);
	} else
	    waddstr (_wmsg, ".\n");
    }
    return card;
}

static bool gofish (enum EPlayer player, unsigned askedfor)
{
    wattr_on (_wmsg, A_BOLD, NULL);
    printplayer (OTHER(player));
    waddstr (_wmsg, "say \"GO FISH!\"\n");
    wattr_off (_wmsg, A_BOLD, NULL);
    if (askedfor == drawcard (player)) {
	printplayer (player);
	waddstr (_wmsg, "drew the guess and get to ask again.\n");
	return true;
    }
    return false;
}

static void goodmove (enum EPlayer player, unsigned m)
{
    printplayer (OTHER(player));
    wprintw (_wmsg, "have %u %s%s.\n", _hand[OTHER(player)][m], c_CardNames[m], _hand[OTHER(player)][m] == 1 ? "" : "'s");

    _hand[player][m] += _hand[OTHER(player)][m];
    _hand[OTHER(player)][m] = 0;

    if (_hand[player][m] == SUITS) {
	printplayer (player);
	wprintw (_wmsg, "made a book of %s's!\n", c_CardNames[m]);
	chkwinner (player);
    }
    chkwinner (OTHER(player));

    printplayer (player);
    waddstr (_wmsg, "get another guess!\n");
}

static void chkwinner (enum EPlayer player)
{
    for (unsigned i = 0; i < RANKS; ++i)
	if (_hand[player][i] > 0 && _hand[player][i] < SUITS)
	    return;
    draw_panel();
    printplayer (player);
    waddstr (_wmsg, "don't have any more cards!\n\nMy books:");
    unsigned cb = countbooks (COMPUTER);
    waddstr (_wmsg, "Your books:");
    unsigned ub = countbooks (USER);
    wprintw (_wmsg, "\nI have %d, you have %d.\n", cb, ub);
    if (ub > cb) {
	waddstr (_wmsg, "\nYou win!!!\n");
	if (!nrand(64))
	    waddstr (_wmsg, "Cheater, cheater, pumpkin eater!\n");
    } else if (cb > ub) {
	waddstr (_wmsg, "\nI win!!!\n");
	if (!nrand(64))
	    waddstr (_wmsg, "Hah! Stupid peasant!\n");
    } else
	waddstr (_wmsg, "\nTie!\n");
    wgetch (_wmsg);
    exit (EXIT_SUCCESS);
}

static void printplayer (enum EPlayer player)
{
    wprintw (_wmsg, "%s ", c_PlayerName[player]);
}

static unsigned countbooks (enum EPlayer player)
{
    unsigned count = 0;
    for (unsigned i = 0; i < RANKS; ++i) {
	if (_hand[player][i] == SUITS) {
	    ++count;
	    wprintw (_wmsg, " %s", c_CardNames[i]);
	}
    }
    if (!count)
	waddstr (_wmsg, " none");
    waddch (_wmsg, '\n');
    return count;
}

static void instructions (void)
{
    waddstr (_wmsg, 
	"\nThis is the traditional children's card game \"Go Fish\". We each get\n"
	"seven cards, and the rest of the deck is kept to be drawn from later. The\n"
	"object of the game is to collect \"books\", or all of the cards of a single\n"
	"value. For example, getting four 2's would give you a \"book of 2's\".\n"
	"\n"
	"We take turns asking each other for cards, but you can only ask me for\n"
	"a card value you already have. If I have any, I must give them to you,\n"
	"and you can keep asking. If you ask me for a card value I don't have,\n"
	"then I'll tell you to \"Go Fish!\", which means that you draw a card from\n"
	"the deck. If you draw the card you asked me for, you get to keep asking\n"
	"me for cards. If not, then it's my turn. The game stops when one of us\n"
	"runs out of cards, and the winner is the one with the most books.\n"
	"\n"
	"You ask for a card by typing its name (a234567891jqk), type x to quit,\n\n"
    );
}
